### PR TITLE
BarChart: Support numeric wide/multi out-of-the-box

### LINKED ISSUE #36712: BarChart: Support Prometheus Instant Queries (Labeled Data)
**What would you like to be added**:
Support for Barchart display of "instant" Prometheus queries - in which the identifier of each item in the bar graph is the set of labels and not string columns.

Currently Get:
image

**Why is this needed**:
Popular data source and response format (note SSE shares similar response format for number data - except no time field), so having to use the Labels to Fields transform is cumbersome.

Note: This also relates to finding a numeric data format to support Grafana Managed alerts with SSE https://github.com/grafana/grafana/issues/35663 to enable https://grafana.com/docs/grafana/latest/alerting/unified-alerting/grafana-managed-numeric-rule/ with non-SQL datasources, and more generally this [internal design doc](https://docs.google.com/document/d/1pR9g_LQfjl5nCCIZzjV3pJGVCKfB9MtqNehp6ZJVEMI/edit) on data types

### PR DESCRIPTION
**Pull request by ryantxu**
**What is this feature?**

Similar to https://github.com/grafana/grafana/pull/62732, this PR makes the bar chart render reasonable results for all `numeric-*` data frame types.  https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/contract_docs/numeric.md

image

**Why do we need this feature?**

Currently numeric types require additional processing to show as bar charts.  EG:  
image

**Which issue(s) does this PR fix?**:

With additional work in the prometheus data source, this could fix https://github.com/grafana/grafana/issues/36712

### CODE REVIEW

**Comment by ryantxu** [comment]
@kylebrandt -- if this looks reasonable, I'll add some more tests

**Comment by kylebrandt** [comment]
With the following Long Numeric Frame:

<details>
<summary>Long Frames JSON (one frame)</summary>

```json
[
  {
    "schema": {
      "meta": {
        "type": "numeric-long"
      },
      "fields": [
        {
          "type": "number",
          "name": "avgSlothCount",
          "typeInfo": {
            "frame": "float64"
          }
        },
        {
          "type": "number",
          "name": "avgSleepHoursPerSlothPerDay",
          "typeInfo": {
            "frame": "float64"
          }
        },
        {
          "type": "string",
          "name": "city",
          "typeInfo": {
            "frame": "string"
          }
        }
      ]
    },
    "data": {
      "values": [
        [4, 7.5],
        [23.5, 23.2],
        ["LGA", "MIA"]
      ]
    }
  }
]
```
</details>

I get the following:
image

However with an equivalent numerc wide:

<details>
<summary>Wide Frames JSON (also one frame)</summary>

```json
[
  {
    "schema": {
    "meta": {
        "type": "numeric-wide"
    },
    "fields": [
        {
          "type": "number",
          "name": "avgSlothCount",
          "typeInfo": {
              "frame": "float64"
          },
          "labels": {
              "city": "LGA"
          }
        },
        {
          "type": "number",
          "name": "avgSlothCount",
          "typeInfo": {
              "frame": "float64"
          },
          "labels": {
              "city": "MIA"
          }
        },
        {
          "type": "number",
          "name": "avgSleepHoursPerSlothPerDay",
          "typeInfo": {
              "frame": "float64"
          },
          "labels": {
              "city": "LGA"
          }
        },
        {
          "type": "number",
          "name": "avgSleepHoursPerSlothPerDay",
          "typeInfo": {
              "frame": "float64"
          },
          "labels": {
              "city": "MIA"
          }
        }
      ]
    },
    "data": {
      "values": [
        [
          4
        ],
        [
          7.5
        ],
        [
          23.5
        ],
        [
          23.2
        ]
      ]
    }
  }
]
```
</details>

I get the following:
image

Differences:

- Each unique "name" is displayed in the legend next to the color (Aside: I think the long example should have 2 different colors in it (and wide/multi as well))
- The bars are grouped spatially by matching dimenstions/labels in long, but not here.

**Comment by kylebrandt** [comment]
Hmm... seems to be a current limitation of the X axis in long that it can only be one string field, and not multiple string fields combined. So I guess there are improvements to make with NumericLong as well (color, more than one dimension grouping)

File: public/app/plugins/panel/barchart/utils.test.ts
Comment (ryantxu) [inline public/app/plugins/panel/barchart/utils.test.ts:169]: needed to change this because MutableDataFrame will change to numbers if not specified

File: public/app/plugins/panel/barchart/utils.ts
Comment (ryantxu) [inline public/app/plugins/panel/barchart/utils.ts:415]: most changes below here are moving existing code into a function.

File: public/app/plugins/panel/barchart/utils.ts
Comment (ryantxu) [inline public/app/plugins/panel/barchart/utils.ts:583]: this avoids showing "frame is missing a string field" if the value can be converted to numericLong

**Comment by kylebrandt** [comment]
Generally Working for me, three cases to address (or make an issue of and merge now), No Data, Empty Items, and a Rendering (Long Conversion?) Issue:

## No Data

A No Data frame has no fields (so we can still pass metadata). https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/contract_docs/contract.md#no-data-and-empty

<details>
    <summary>Numeric Multi Frames JSON (NO Data, Multi)</summary>
    
```json
[
  {
    "schema": {
    "meta": {
        "type": "numeric-multi",
        "typeVersion": [0, 1]
    },
    "fields": []
    },
    "data": {
      "values": []
    }
  }
]
```

</details>

Looks like the following:
image

But I think it should say "No Data" without the Value item in the legend. I also think NumericLong should be the same, but it currently shows:

image

Also, NumericWide should be have the same.

## Empty

Empty has fields, but of zero length. https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/contract_docs/contract.md#no-data-and-empty

### Long

<details>
    <summary>NumericLong Frames JSON (note: should values have sub arrays? seems to not make difference)</summary>
    
```json
[
  {
    "schema": {
      "meta": {
        "type": "numeric-long",
        "typeVersion": [0, 1]
      },
      "fields": [
        {
          "type": "number",
          "name": "avgSlothCount",
          "typeInfo": {
            "frame": "float64"
          }
        },
        {
          "type": "number",
          "name": "avgSleepHoursPerSlothPerDay",
          "typeInfo": {
            "frame": "float64"
          }
        },
        {
            "type": "string",
            "name": "city",
            "typeInfo": {
              "frame": "string"
            }
        }
      ]
    },
    "data": {
      "values": []
    }
  }
]
```

</details>

This Displays the value fields but not the dimensions:

image

I think the dimensions should be displayed but no bars rendered.

Wide and Multi do not show the values or dimensions (and I think it would be better if they did):

<details>
    <summary>Example Frames JSON Wide and Empty</summary>
    
```json
[
  {
    "schema": {
      "meta": {
        "type": "numeric-wide",
        "typeVersion": [0, 1]
      },
      "fields": [
        {
          "type": "number",
          "name": "avgSlothCount",
          "typeInfo": {
            "frame": "float64"
          },
          "labels": {
            "city": "LGA"
          }
        },
        {
          "type": "number",
          "name": "avgSlothCount",
          "typeInfo": {
            "frame": "float64"
          },
          "labels": {
            "city": "MIA"
          }
        }
      ]
    },
    "data": {
      "values": [[], []]
    }
  }
]

```

</details>

image


## Visual of single grouped items

With numeric-wide or multi, the bars are off to the side (I'm guessing this in conversion to long, perhaps `undefined` here will help in the specific case of the FE).

<details>
    <summary>Wide Frames JSON that renders funny</summary>
    
```json
[
  {
    "schema": {
      "meta": {
        "type": "numeric-wide",
        "typeVersion": [0, 1]
      },
      "fields": [
        {
          "type": "number",
          "name": "avgSlothCount",
          "typeInfo": {
            "frame": "float64"
          },
          "labels": {
            "city": "LGA"
          }
        },
        {
          "type": "number",
          "name": "avgSleepHoursPerSlothPerDay",
          "typeInfo": {
            "frame": "float64"
          },
          "labels": {
            "city": "MIA"
          }
        }
      ]
    },
    "data": {
      "values": [[4], [7.5]]
    }
  }
]

```

</details>

image

So the first 2 are specific to dataplane, last one seems minor. But they can all go into an issue if you want to merge now and treat these separately.

**Review by kylebrandt — APPROVED** [review]
(Not a code review)

Generally working well for me. Left some examples of some of the more edge cases in my last comment. Leave it to you if you want to address here, or have those comments moved to a new issue after merging this.

**Pull request closed without merging**
